tools/integration-tests/tester/framework/network.go

```diff
@@ -0,0 +1,179 @@
+	if err != nil {
+		panic(err)
+	}
+
+	// the tester container needs to join the Docker network in order to communicate with the peers
+	tester.ConnectToNetwork(resp.ID)
+
+	return &Network{
+		id:           resp.ID,
+		name:         name,
+		tester:       tester,
+		dockerClient: dockerClient,
+	}
+}
+
+// createEntryNode creates the network's entry node.
+func (n *Network) createEntryNode() {
+	// create identity
+	publicKey, privateKey, err := hive_ed25519.GenerateKey()
+	if err != nil {
+		panic(err)
+	}
+
+	n.entryNodeIdentity = identity.New(publicKey)
+	seed := base64.StdEncoding.EncodeToString(ed25519.PrivateKey(privateKey.Bytes()).Seed())
+
+	// create entry node container
+	n.entryNode = NewDockerContainer(n.dockerClient)
+	n.entryNode.CreateGoShimmerEntryNode(n.namePrefix(containerNameEntryNode), seed)
+	n.entryNode.ConnectToNetwork(n.id)
+	n.entryNode.Start()
+}
+
+// CreatePeer creates a new peer/GoShimmer node in the network and returns it.
+func (n *Network) CreatePeer() *Peer {
+	name := n.namePrefix(fmt.Sprintf("%s%d", containerNameReplica, len(n.peers)))
+
+	// create identity
+	publicKey, privateKey, err := hive_ed25519.GenerateKey()
+	if err != nil {
+		panic(err)
+	}
+	seed := base64.StdEncoding.EncodeToString(ed25519.PrivateKey(privateKey.Bytes()).Seed())
+
+	// create Docker container
+	container := NewDockerContainer(n.dockerClient)
+	container.CreateGoShimmerPeer(name, seed, n.namePrefix(containerNameEntryNode), n.entryNodePublicKey())
+	container.ConnectToNetwork(n.id)
+	container.Start()
+
+	peer := newPeer(name, identity.New(publicKey), container)
+	n.peers = append(n.peers, peer)
+	return peer
+}
+
+// Shutdown creates logs and removes network and containers.
+// Should always be called when a network is not needed anymore!
+func (n *Network) Shutdown() {
+	// stop containers
+	n.entryNode.Stop()
+	for _, p := range n.peers {
+		p.Stop()
+	}
+
+	// retrieve logs
+	createLogFile(n.namePrefix(containerNameEntryNode), n.entryNode.Logs())
+	for _, p := range n.peers {
+		createLogFile(p.name, p.Logs())
+	}
+
+	// remove containers
+	n.entryNode.Remove()
+	for _, p := range n.peers {
+		p.Remove()
+	}
+
+	// disconnect tester from network otherwise the network can't be removed
+	n.tester.DisconnectFromNetwork(n.id)
+
+	// remove network
+	err := n.dockerClient.NetworkRemove(context.Background(), n.id)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// WaitForAutopeering waits until all peers have reached the minimum amount of neighbors.
+// Panics if this minimum is not reached after autopeeringMaxTries.
+func (n *Network) WaitForAutopeering(minimumNeighbors int) {
+	fmt.Printf("Waiting for autopeering...\n")
+	defer fmt.Printf("Waiting for autopeering... done\n")
+
+	maxTries := autopeeringMaxTries
+	for maxTries > 0 {
+
+		for _, p := range n.peers {
+			if resp, err := p.GetNeighbors(false); err != nil {
+				fmt.Printf("request error: %v\n", err)
```

Comment:
Is it expected behavior that the request fails? And if yes, is it correct to ignore this peer?
Wouldn't it maybe more consistent to cancel WaitForAutopeering with an error in this case

Reply:
It can happen that a container and its GoShimmer instance is not yet fully started once this is called. This obviously depends on the underlying system and I don't think we have another way for checking this except maybe [Docker's healthcheck feature](https://docs.docker.com/compose/compose-file/#healthcheck). But this would make it unnecessarily complicated.

Reply:
OK, then let's keep it like this!